src/main/java/net/wurstclient/hacks/AutoCraftHack.java

```diff
@@ -0,0 +1,440 @@
+package net.wurstclient.hacks;
+
+import net.minecraft.client.network.ClientPlayerInteractionManager;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.recipe.Ingredient;
+import net.minecraft.recipe.Recipe;
+import net.minecraft.recipe.RecipeManager;
+import net.minecraft.recipe.RecipeType;
+import net.minecraft.screen.CraftingScreenHandler;
+import net.minecraft.screen.slot.SlotActionType;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+import net.wurstclient.Category;
+import net.wurstclient.events.UpdateListener;
+import net.wurstclient.hack.Hack;
+
+import java.util.*;
+import java.util.stream.Stream;
+
+public class AutoCraftHack extends Hack implements UpdateListener {
+
+    private HashMap<Identifier, List<Recipe<?>>> recipeMap = null;
+
+    private List<CraftingQueueEntry> craftingQueue = new ArrayList<>();
+
+    private boolean isCurrentlyCrafting = false;
+
+    public AutoCraftHack() {
+        super("AutoCraft");
+        setCategory(Category.ITEMS);
+    }
+
+    private void initRecipeMap() {
+        recipeMap = new HashMap<>();
+        RecipeManager recipeManager = MC.world.getRecipeManager();
+        Stream<Identifier> keys = recipeManager.keys();
+        Iterator<Identifier> iter = keys.iterator();
+        while (iter.hasNext()) {
+            Identifier id = iter.next();
+            Optional<? extends Recipe<?>> optionalRecipe = recipeManager.get(id);
+            if (!optionalRecipe.isPresent())
+                continue;
+            Recipe recipe = optionalRecipe.get();
+            RecipeType<?> recipeType = recipe.getType();
+            if (recipeType != RecipeType.CRAFTING)
+                continue;
+            String path = id.getPath();
+            if (path.contains("_from_")) {
+                String[] components = path.split("_from_");
+                String sourceId = components[0];
+                Identifier baseId = new Identifier(id.getNamespace(), sourceId);
+                if (!recipeMap.containsKey(baseId))
+                    recipeMap.put(baseId, new ArrayList<Recipe<?>>());
+                recipeMap.get(baseId).add(recipe);
+            }
+            else {
+                if (!recipeMap.containsKey(id))
+                    recipeMap.put(id, new ArrayList<Recipe<?>>());
+                recipeMap.get(id).add(recipe);
+            }
+        }
+    }
+
+    private abstract class StorageQuery {
+        public abstract int getAvailable(Item item);
+        public abstract HashMap<Item, Integer> getAvailabilityMap();
+        public abstract void acquire(Item item, int count);

```

Comment:
To add functionality to this method, we can set a StoragreQuery that detects a storage container nearby and opens it, and extracts the necessary items. Another option, which is more complicated, is to create a package of methods to get an item. This is complicated because it would require pathfinding ai and a class for every single item.

Reply:
For the latter approach, I'm thinking of interfacing with baritone for pathfinding. This would of course require the user to have baritone installed alongside wurst, but it seems worthwhile given how well baritone does pathfinding. There would need to be some balance between looking for items in storage containers and harvesting them from the world. Perhaps this could involve some spatial logic (if nearby chests are empty, then probably this one will be too), and some efficiency heuristics based on the player's distance from the storage container versus their distance from the blocks in the world and the expected mining time.